### PR TITLE
Refine generic link text linter logic

### PR DESCRIPTION
**Pull request by khiga8**
Relates to: https://github.com/github/accessibility/issues/1291

@github/accessibility-reviewers before you review, please check the discussion in the linked issue :) 

## Changes
This refines the generic linter text introduced in https://github.com/github/erblint-github/pull/32 based on internal discussions.

Currently, this linter will flag a link with generic text even if it has a descriptive accessible name set by `aria-labelledby` or `aria-label`. This is a _sometimes_ acceptable technique even if it may not be the most preferable, so we don't want to raise a flag because it could be a false positive. 

This updates the linter to not evaluate links with generic text that have ARIA attributes since linters should never raise false positives. One nuance we discussed is that when using ARIA to set a more descriptive accessible name, one _must_ ensure that the visible text is contained in the accessible name output. This is because this would be a violation of [Success Criterion 2.5.3: Label in Name](https://www.w3.org/WAI/WCAG21/Understanding/label-in-name). 

If the linter is able to access the `aria-label` string and _can accurately_ determine that the `aria-label` text does not contain the visible text at all, it **will** raise a flag. Notably, because this is static code analysis, we have limitations when it comes to evaluating the actual runtime values of the accessible name output, so this flag won't always be raised.

### Examples of accessible name that ERB lint can and cannot evaluate

ERB lint can evaluate the accessible name when `aria-label` is a string:
```
<a href="github.com" aria-label='Learn more about GitHub'>Learn more</a>

```

ERB lint cannot evaluate the accessible name when `aria-labelledby` so we skip:
```
<a href="github.com" aria-labelledby='element123'>Learn more</a>

```

ERB lint cannot evaluate the accessible name since it's a variable so we skip:
```
<a href="github.com" aria-label="<%= some_variable %>">Learn more</a>

```

ERB lint cannot accurately evaluate the accessible name of `link_to` helpers with `aria-label` or `aria-labelledby` attributes so we skip completely:
```
<%= link_to "learn more", billing_path, "aria-label": "learn #{@some_variable}" %>
# The aria-label output by ERB is simply `learn` which is inaccurate
```
Even with these limitations around not being able to comprehensively evaluate links if they have ARIA label attributes set, this linter should catch quite a number of issues.

I've added a note in the doc that further informs these changes. The code is a bit complex so I added comments and additional tests which should help. 

## References
GitHub Staff only - [Discussion on nuanced uses of ARIA to provide more descriptive text](https://github.com/github/accessibility/issues/1291#issuecomment-1159088348)

### CODE REVIEW

File: lib/erblint-github/linters/github/accessibility/avoid_generic_link_text_counter.rb
Comment (khiga8) [inline lib/erblint-github/linters/github/accessibility/avoid_generic_link_text_counter.rb:47]: We skip flagging the link if it has `aria-labelledby` or an aria-label that is set to a variable like:

```.erb
<a href="github.com" aria-label="<%= some_variable%>" >
```
since it may be acceptable usecase and we cannot evaluate with linter.

File: docs/rules/accessibility/avoid-generic-link-text-counter.md
Comment (andrialexandrou) [inline docs/rules/accessibility/avoid-generic-link-text-counter.md:30]: If you _must_ use ARIA to replace the visible link text, include the visible text at the beginning.

For example, on a pricing plans page, the following are good:
- Visible text: `Learn more`
- Accessible label: `Learn more about GitHub pricing plans`

Accessible ✅
```html
<a href="..." aria-label="Learn more about GitHub pricing plans">Learn more</a>
```

Inaccessible 🚫
```html
<a href="..." aria-label="GitHub pricing plans">Learn more</a>
```

Including the visible text in the ARIA label satisfies [SC 2.5.3: Label in Name](https://www.w3.org/WAI/WCAG21/Understanding/label-in-name.html).

#### False negatives

Caution: because of the restrictions of static code analysis, we may not catch all violations.

Please perform browser tests and spot checks:
- when `aria-label` is set dynamically
- when using `aria-labelledby`

**Review by andrialexandrou — APPROVED** [review]
Ship to learn ✨

Nonblocking:

I put a suggested change to the documentation in a comment. Bc of code blocks and suggestions not playing nice together, I'll leave it to you to add it in a commit or not. I don't want to push to your branch w/o approval.

**Review by owenniblock — APPROVED** [review]

**Pull request merged**
